### PR TITLE
#68 - Auth testing & playwright improvements

### PR DESCRIPTION
**Pull request by archae0pteryx**
- Added script for dev friendly playwright testing.
- Updated github action to use postgres user to fix a swallowed action error
- Add `DISABLE_EMAIL` env var to skip trying to send email locally
- Added a cascade delete for users' membership for deleting users during e2e tests
- Added auto login for auth user and basic user in playwright
- Added user entity for user manipulation in e2e tests
- Implemented POM (Page Object Model) pattern for playwright tests

### CODE REVIEW

File: web/src/components/Profile/EditProfile/EditProfile.tsx [hunk 1]
```diff
@@ -9,7 +9,7 @@ import {
 
 const EditProfile = ({ error, loading, profile, onSave }) => {
   return (
-    <div className="rw-form-wrapper">
+    <div id="edit-profile" className="rw-form-wrapper">
```
Comment (craineum) [inline web/src/components/Profile/EditProfile/EditProfile.tsx:12]: ❓ Can playwright use the `data-testid` attribute instead?

File: web/tests/signup.spec.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    await expect(signupTitle).toBeVisible()
+    await expect(signupTitle).toHaveText('Signup')
+
+    await page.locator('input[name="username"]').click()
+    await page.locator('input[name="username"]').fill(MOCK_USER.email)
+    await page.locator('input[name="password"]').click()
+    await page.locator('input[name="password"]').fill('example')
+
+    await page.locator('button:has-text("Sign Up")').click()
+    await page.waitForURL('/')
```
Comment (craineum) [inline web/tests/signup.spec.ts:36]: ❓ Should we make sure there is a `logout` option or some kind of toast?

File: web/tests/storage/usersState.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export const ADMIN_USER_STATE = './adminUser-pw.json'
```
Comment (craineum) [inline web/tests/storage/usersState.ts:1]: ❓ I don't see this used anywhere, is it playwright magic that is doing something with these?

**Review by craineum — COMMENTED** [review]
Looking Good, some questions.

File: .env.defaults [hunk 4]
```diff
@@ -1,6 +1,11 @@
 DATABASE_URL=postgresql://postgres:development@localhost:5432/redwood_dev
 TEST_DATABASE_URL=postgresql://postgres:test@localhost:5433/redwood_test
 
+# DATABASE_URL=postgresql://postgres:test@localhost:5433/redwood_test
```
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply .env.defaults:4]: 🧹

File: api/db/migrations/20221110212458_init/migration.sql [hunk 5]
```diff
@@ -11,8 +11,9 @@ CREATE TABLE "User" (
     "salt" TEXT NOT NULL,
     "resetToken" TEXT,
     "resetTokenExpiresAt" TIMESTAMP(3),
-    "updatedAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
-    "createdAt" TIMESTAMP(3) NOT NULL,
+    "updatedAt" TIMESTAMP(3) NOT NULL,
```
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply api/db/migrations/20221110212458_init/migration.sql:14]: The migrations are totally reset. I should have made a note to push the new schema locally. `yarn rw prisma db push` should update your local dev. If it still has an issue you might need to just remove the data volumes `docker compose down -v`

File: scripts/seed.ts [hunk 6]
```diff
@@ -81,6 +81,11 @@ async function _createRole(role) {
 
 export default async () => {
   try {
+    const hasUsers = await db.user.findMany()
+    if (hasUsers.length) {
+      console.log('Has database alread been seeded?')
```
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply scripts/seed.ts:86]: 🧹

File: web/src/components/Profile/EditProfile/EditProfile.tsx [hunk 7]
```diff
@@ -79,6 +79,7 @@ const EditProfile = ({ error, loading, profile, onSave }) => {
 
         <TextField
           name="email"
+          disabled
```
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply web/src/components/Profile/EditProfile/EditProfile.tsx:82]: It breaks for some reason at the moment. We thought it better to create a separate ticket. I forgot to make a note of it so I will create another ticket to figure out how to make this work.

File: web/src/pages/ForgotPasswordPage/ForgotPasswordPage.tsx [hunk 8]
```diff
@@ -15,13 +15,13 @@ const ForgotPasswordPage = () => {
     }
   }, [isAuthenticated])
 
-  const emailRef = useRef<HTMLInputElement>()
+  const usernameRef = useRef<HTMLInputElement>()
```
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply web/src/pages/ForgotPasswordPage/ForgotPasswordPage.tsx:18]: I asked the same question. @nataliadiaz2 said she was trying to match the API from the password reset (if i remember correctly). I agree that it should stay email and that the other end should match. Want us to tackle this and the email edit in this PR while we're at it or make separate tickets? I was just feeling like this one was creeping outta scope.

File: web/src/components/Profile/EditProfile/EditProfile.tsx (hunk 1 above)
Comment (archae0pteryx) [reply web/src/components/Profile/EditProfile/EditProfile.tsx:12]: =( my fast / basic search showed, no... And the id's don't work on redwood's form elements. I suspect the props aren't being passed all the way through to the raw element which makes playwright not be able to select it. Here it works cause it's right on the html element... need to figure out a good selection pattern that's not a class or text... Maybe another ticket for this?

File: web/tests/signup.spec.ts (hunk 2 above)
Comment (archae0pteryx) [reply web/tests/signup.spec.ts:36]: yup. there's a few cases not covered.

File: web/tests/storage/usersState.ts (hunk 3 above)
Comment (archae0pteryx) [reply web/tests/storage/usersState.ts:1]: They're not used *yet* we set it up to automatically have an admin user or a basic user be authed in the tests (as opposed to manually logging someone in like we were). The next round of tests / backfill will be using all the great things we've learned!

File: api/db/migrations/20221110212458_init/migration.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (craineum) [reply api/db/migrations/20221110212458_init/migration.sql:14]: I don't think we can keep doing this, If outside developers are using this template at this point.

File: web/src/pages/ForgotPasswordPage/ForgotPasswordPage.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (craineum) [reply web/src/pages/ForgotPasswordPage/ForgotPasswordPage.tsx:18]: Yeah, I agree this should be a separate ticket.

File: web/src/components/Profile/EditProfile/EditProfile.tsx (hunk 1 above)
Comment (craineum) [reply web/src/components/Profile/EditProfile/EditProfile.tsx:12]: Yeah, good to leave this and open another ticket.

File: api/db/migrations/20221110212458_init/migration.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (archae0pteryx) [reply api/db/migrations/20221110212458_init/migration.sql:14]: You're totally right... It just pains me to be tracking the migrations created during active development (like the ones we made for this PR). My reasoning in this particular case is that our "production" db's aren't prod (and potentially moving anyway) and cleaning this up locally (i don't think) is a huge problem.

**Pull request merged**